### PR TITLE
New IA: Calculator

### PR DESCRIPTION
**Pull request by sevastos**
#### Math calculator - WIP
A math calculator with GUI.

Screenshot

#### State
- Demo: To test put a math expression followed by an equal sign (`=`) e.x: `2+5^3=`
- Trigger: Uses a regex for trigger searching for a `=` at the end of the query.
 - [ ]Revisit and finish properly triggering (to not conflict with original goodie?)
- JS: Most of the foundation is in place, but restructuring is planned.
- UI: Most CSS is done, except:
  - [ ] Dark theme support
  - [ ] Extensive cross platform (IE) testing
  - [ ] UI transitions
  - [ ] More testing on client configurations
- UI/JS: Still todo:
  - [ ] History log:
    - [ ] more/less toggle
    - [ ] replay of past formulas
    - [ ] misc ideas
  - [ ] Change angle toggle with label and input(type=checkbox)
  - [ ] i18n
- Formula: Basic operations work, but a lot of the functionality is disabled or is not fully developed. Complicated formulas may not work yet. A lot of progress is to happen in the following days to the point of being complete.
  - [ ] Deleting/Backspace (this is quite crucial; WIP to cover all cases)
  - [ ] Parentheses
  - [X] Basic math ( + - / * )
  - [ ] Trigonometric functions (work with rads, yet to develop to respect the angle unit option
  - [X] Powers, ExpNotation
  - [x] Roots
  - [ ] Constants (any use of constants is not developed yet)
  - [ ] log/ln: ln still to do
  - [X] Factorial 
  - [ ] Flawless navigation :star2:
  - [ ] Flawless complexity :star2: 
- Accessibility: Basic work is done, but there are still obvious things still to do (WAI etc.).
- There is a features backlog, if room/needed
- Refactors are imminent :smiling_imp: 

##### `Calc`
The idea is to have `calc` normalize any user input (UI buttons, keys, clicks) and then feed the "intentions" to the active `calc.formula` (`Formula`)
##### `Formula`
`Formula` handles the presentation and calculating part. It's stores the math expressions in a structured way along with a virtual cursor which we could show on the UI.
This allows for much flexibility on what we can do with the formula ;)

/cc @bsstoner @nilnilnil

### CODE REVIEW

**Comment by mattr555** [comment]
Yay! :tada: (just commenting so I can follow updates)

**Comment by javathunderman** [comment]
:cake: Looking forward to this!

**Comment by chrismorast** [comment]
Looking forward to this as well!

**Comment by jagtalon** [comment]
This is sick! :neckbeard:

**Comment by sevastos** [comment]
Cheers, btw I didn't do the design, but whoever did it, they did really awesome work.

**Comment by bsstoner** [comment]
Does anyone have opinions on using Math.js (http://mathjs.org/)? It's kind of big, but maybe it's not a big deal if we could load it async after the reset of the SERP loads? Is there enough functionality in Math.js to make it worth using or should we just build only what we need here in the IA?

**Comment by jagtalon** [comment]
@bsstoner I think building it out takes time, and we've found a lot of errors while making https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Calculator.pm. If this works and it's been tried and tested, I'm all for it. :)

**Comment by bsstoner** [comment]
@moollaza and I talked about it a bit. I think what Sev wrote here works for now, no need to complicate things with Math.js yet. We can re-evaluate later if there's a clearer need for it.

This is really looking great though, excited to see it live!

I know eventually it will need to be reconciled with the existing Goodie, but for now what do you think about finishing off the remaining issues and launching this in some limited trigger space (i.e. "Calculator" or "Interactive Calculator")? Then as phase 2 we can figure out how to merge it with the existing Goodie (or other backend service) so it shows up for all calculation queries?

**Comment by jagtalon** [comment]
@bsstoner having unit tests for the IAs would work well here! The tests in Calculator.pm helped us out a lot when anyone changed the code. It's gotten quite big now https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/t/Calculator.t

**Comment by javathunderman** [comment]
ping @sevastos @jagtalon @bsstoner Any updates?

**Comment by sevastos** [comment]
@javathunderman I'm back in the game next week, unless someone else is working on it or plan changed.

**Comment by bsstoner** [comment]
Sweet!

**Comment by javathunderman** [comment]
Awesome!

File: share/spice/calculator/calculator.js [hunk 1]
```diff
@@ -0,0 +1,1172 @@
...
+        };
+    }
+
+    // Spice
+    env.ddg_spice_calculator = function(){
+        clog('[calc] Spice inited. Query:', DDG.get_query());
+
+        // Render the response
+        Spice.add({
+            id: 'calculator',
```
Comment (russellholt) [inline share/spice/calculator/calculator.js]: needs to be unique. we already have an IA with id 'calculator'

**Comment by russellholt** [comment]
This needs to verify in perl that the query is a math expression before it gets to the front end.

File: share/spice/calculator/calculator.js (hunk 1 above)
Comment (sevastos) [inline share/spice/calculator/calculator.js]: ok thx, will refactor to 'calculator_gui' unless there's something better.

**Comment by sevastos** [comment]
I was focusing on the UI implementation at first and only did a basic scaffolding on the other parts. Will get back to this. Probably steal and adapt a lot from the calculator goodie.

**Comment by supakeith** [comment]
Yes!

**Comment by javathunderman** [comment]
ping @sevastos Are you interested in finishing this one up?

**Comment by jagtalon** [comment]
@sevastos ping!

**Comment by zekiel** [comment]
@sevastos  if you get around to this, please also change the name:ID to match the id here: https://duck.co/ia/view/interactive_calculator     

it should be:    interactive_calculator

**Comment by moollaza** [comment]
@sevastos I'm closing this one for now, but if you'd like to finish this up, we'll gladly open it back up for you!

**Comment by JamyGolden** [comment]
I'd like to claim this and finish it up.

**Comment by sevastos** [comment]
@JamyGolden :+1: thanks, you're more than welcome by me. Do whatever you like, refactoring / from scratch anything you want.

**Comment by GuiltyDolphin** [comment]
@sevastos @JamyGolden Just so you know, I've been working on this over on duckduckgo/zeroclickinfo-goodies#2189.

Instead of using any JS to do calculations, it sends requests to the Perl back-end (which is in a separate PR - duckduckgo/zeroclickinfo-goodies#1927) which does all the calculations. I've also changed up most of the JS (although I'm still using mostly the same CSS and Handlebars). 

I'm happy to discuss design and functionality (or anything else!) over on that PR, which is where I will be posting progress updates.

/cc @bsstoner @moollaza

**Pull request closed without merging**